### ISSUE TITLE
'the parser seems stuck' when expanding macros

### ISSUE THREAD

**Issue by bjorn3** [rust-analyzer/rust-analyzer#6100]
This happened while working on cg_clif. I can't reproduce it.

<details><summary>Backtrace</summary>

```
thread '<unnamed>' panicked at 'the parser seems stuck', crates/parser/src/parser.rs:51:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:217
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:526
  11: std::panicking::begin_panic
  12: parser::grammar::expressions::atom::atom_expr
  13: parser::grammar::expressions::lhs
  14: parser::grammar::expressions::expr_bp
  15: parser::grammar::expressions::lhs
  16: parser::grammar::expressions::expr_bp
  17: parser::grammar::expressions::atom::atom_expr
  18: parser::grammar::expressions::lhs
  19: parser::grammar::expressions::expr_bp
  20: parser::grammar::expressions::atom::atom_expr
  21: parser::grammar::expressions::lhs
  22: parser::grammar::expressions::expr_bp
  23: parser::grammar::expressions::stmt
  24: parser::grammar::expressions::expr_block_contents
  25: parser::grammar::expressions::atom::block_expr_unchecked
  26: parser::grammar::expressions::atom::atom_expr
  27: parser::grammar::expressions::lhs
  28: parser::grammar::expressions::expr_bp
  29: parser::grammar::fragments::expr
  30: parser::parse_fragment
  31: mbe::syntax_bridge::token_tree_to_syntax_node
  32: hir_expand::db::parse_macro_with_arg
  33: salsa::runtime::Runtime::execute_query_implementation
  34: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  35: salsa::derived::slot::Slot<Q,MP>::read
  36: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::try_fetch
  37: <DB as hir_expand::db::AstDatabase>::parse_macro::__shim
  38: <DB as hir_expand::db::AstDatabase>::parse_macro
  39: hir_expand::db::parse_or_expand
  40: <DB as hir_expand::db::AstDatabase>::parse_or_expand
  41: hir_def::body::Expander::enter_expand
  42: hir_def::body::lower::ExprCollector::collect_expr
  43: hir_def::body::lower::ExprCollector::collect_block
  44: hir_def::body::lower::ExprCollector::collect_expr
  45: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  46: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
  47: hir_def::body::lower::ExprCollector::collect_block
  48: hir_def::body::lower::ExprCollector::collect_expr
  49: hir_def::body::lower::lower
  50: hir_def::body::Body::body_with_source_map_query
  51: salsa::runtime::Runtime::execute_query_implementation
  52: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  53: salsa::derived::slot::Slot<Q,MP>::read
  54: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::try_fetch
  55: <DB as hir_def::db::DefDatabase>::body_with_source_map::__shim
  56: hir::source_analyzer::SourceAnalyzer::new_for_body
  57: hir::semantics::SemanticsImpl::analyze2
  58: hir::semantics::SemanticsImpl::type_of_pat
  59: ide::inlay_hints::inlay_hints
  60: ide::Analysis::inlay_hints
  61: rust_analyzer::handlers::handle_inlay_hints
  62: <F as threadpool::FnBox>::call_box
```

</details>

**Comment by matklad** [comment]
Feels like a bad issue, this pokes a pretty painful bit of logic: https://github.com/rust-analyzer/rust-analyzer/pull/4680#issuecomment-641346748

**Comment by jonas-schievink** [comment]
Reproduction:

```rust
    assert_eq!(
        Command::decode(b"S9\r").unwrap_err().kind(),
        ErrorKind::
    );
```

**Comment by matklad** [comment]
closed by #7211 it seems?
